File: ios/SecretaryKit/Sources/SecretaryKit/DeviceUnlock/UniffiVaultDeviceSlotPort.swift
import Foundation
import SecretaryDeviceUnlock

/// Real `VaultDeviceSlotPort` over the B.2 uniffi functions. This is the ONLY
/// place that touches the one-shot `DeviceSecretOutput`.
public struct UniffiVaultDeviceSlotPort: VaultDeviceSlotPort {
    public init() {}

    public func addDeviceSlot(vaultPath: Data, password: [UInt8]) throws -> EnrolledSlot {
        do {
            // Scrub the adapter-owned password Data copy on the way out (#364);
            // this runs on the production "Remember this device" enroll path, so
            // a master-password copy must not linger in the heap. Parity with
            // UniffiVaultOpenPort.
            let out = try withZeroizingData(password) { pw in
                try SecretaryKit.addDeviceSlot(folderPath: vaultPath, password: pw)
            }
            // Wipe the one-shot handle on EVERY exit (success, throw, or any
            // future fallible code inserted below), so the secret is never left
            // recoverable in the bridge handle.
            defer { out.deviceSecret.wipe() }
            // takeSecret() is `bytes?` in the UDL → a zeroizable `Data?` (not a boxed list, #261).
            guard var secret = out.deviceSecret.takeSecret() else {
                throw VaultSlotError.other("device secret handle was empty")
            }
            // Zero the transient Data once the port's `[UInt8]` copy is built; the coordinator
            // zeroizes that EnrolledSlot copy after enclave.store.
            defer { secret.resetBytes(in: 0..<secret.count) }
            return EnrolledSlot(deviceUuid: [UInt8](out.deviceUuid), deviceSecret: [UInt8](secret))
        } catch let e as VaultError {
            throw mapVaultError(e)
        }
    }

    public func openWithDeviceSecret(vaultPath: Data, deviceUuid: [UInt8], deviceSecret: [UInt8]) throws -> OpenedVault {
        do {
            // Scrub the adapter-owned device-secret Data copy on the way out
            // (#364). deviceUuid is not secret. Parity with UniffiVaultOpenPort.
            return try withZeroizingData(deviceSecret) { secretData in
                try SecretaryKit.openWithDeviceSecret(
                    folderPath: vaultPath,
                    deviceUuid: Data(deviceUuid),
                    deviceSecret: secretData)
            }
        } catch let e as VaultError {
            throw mapVaultError(e)
        }
    }

    public func removeDeviceSlot(vaultPath: Data, deviceUuid: [UInt8]) throws {
        do {
            try SecretaryKit.removeDeviceSlot(folderPath: vaultPath, deviceUuid: Data(deviceUuid))
        } catch let e as VaultError {
            throw mapVaultError(e)
        }
    }
}

/// Map the uniffi `VaultError` onto the pure `VaultSlotError` mirror.
/// File-private: the device-slot error translation must not be reused to map a
/// `VaultError` from a non-device-slot path through these typed cases.
private func mapVaultError(_ e: VaultError) -> VaultSlotError {
    switch e {
    case .DeviceSlotNotFound:                       return .deviceSlotNotFound
    case .WrongDeviceSecretOrCorrupt:               return .wrongDeviceSecretOrCorrupt
    case .DeviceUuidMismatch(let detail):           return .deviceUuidMismatch(detail)
    case .InvalidArgument(let detail):              return .invalidArgument(detail)
    default:                                        return .other(String(describing: e))
    }
}

File: android/kit/src/main/kotlin/org/secretary/browse/UniffiVaultDeviceSlotPort.kt
package org.secretary.browse

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import uniffi.secretary.DeviceEnrollOutput
import uniffi.secretary.addDeviceSlot as ffiAddDeviceSlot
import uniffi.secretary.removeDeviceSlot as ffiRemoveDeviceSlot

/**
 * The real [VaultDeviceSlotPort] over the generated `add_device_slot` / `remove_device_slot`. Runs on
 * [ioDispatcher] (add_device_slot password-opens the vault → Argon2id). The one-shot
 * `DeviceSecretOutput` is `takeSecret()`-ed once then `wipe()`-d in a `finally` so the bridge retains
 * nothing (mirror of iOS's `defer { out.deviceSecret.wipe() }`). The FFI fns are injectable seams
 * defaulting to the real bindings.
 */
class UniffiVaultDeviceSlotPort(
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
    private val addFn: (ByteArray, ByteArray) -> DeviceEnrollOutput = ::ffiAddDeviceSlot,
    private val removeFn: (ByteArray, ByteArray) -> Unit = ::ffiRemoveDeviceSlot,
) : VaultDeviceSlotPort {
    override suspend fun addDeviceSlot(vaultFolder: String, password: ByteArray): EnrolledSlot =
        withContext(ioDispatcher) {
            mapErrors {
                val out = addFn(vaultFolder.toByteArray(Charsets.UTF_8), password)
                try {
                    // takeSecret() is `bytes?` in the UDL, so uniffi hands back a zeroizable ByteArray? directly —
                    // no intermediate boxed List<UByte> left un-overwritable in the heap (#261). The FFI handle is
                    // wiped in the finally; the coordinator zeroizes this caller-owned ByteArray after enclave.store
                    // (see DeviceUnlockCoordinator.enroll).
                    val secret = out.deviceSecret.takeSecret()
                        ?: throw VaultBrowseError.Failed("device secret handle was empty (already taken?)")
                    EnrolledSlot(out.deviceUuid, secret)
                } finally {
                    out.deviceSecret.wipe()
                }
            }
        }

    override suspend fun removeDeviceSlot(vaultFolder: String, deviceUuid: ByteArray) =
        withContext(ioDispatcher) {
            mapErrors { removeFn(vaultFolder.toByteArray(Charsets.UTF_8), deviceUuid) }
        }
}
